src/ipc/mod.rs

```diff
@@ -69,47 +81,133 @@ impl<T: Ipc> Clone for BackendSender<T> {
         continue_listening: Arc<atomic::AtomicBool>,
-        receive_buf: &'a mut [u8],
-    ) -> Backend<'a, T> {
-        Backend {
-            sock: Rc::new(sock),
+    ) -> Self {
+
+        let mut backends = Vec::new();
+        let mut receivers = Vec::new();
+
+        for sock in socks {
+            let mut backend = SingleBackend::new(sock, Arc::clone(&continue_listening));
+            backends.push(backend.sender());
+
+            let (s,r) = unbounded();
+            receivers.push(r);
+            //sel.recv(&receivers[receivers.len()-1]);
+
+            std::thread::spawn(move || {
+                loop {
+                    s.send(backend.next()).unwrap() // TODO don't unwrap
+                }
+            });
+        }
+
+        let mut sel = Select::new();
+        let recv_ptr = Box::into_raw(Vec::into_boxed_slice(receivers));
+        let recv_slice : &'static [Receiver<Option<Msg>>] = unsafe { &*recv_ptr } ;
+        for r in recv_slice { 
+            sel.recv(r);
+        }
+
+        MultiBackend {
+            last_recvd: Some(0), // TODO need a better solution for this, should be none
             continue_listening,
-            receive_buf,
-            tot_read: 0,
-            read_until: 0,
+            sel,
+            backends,
+            receivers : recv_slice,
+            receivers_ptr: recv_ptr,
         }
     }
+}
+impl<T: Ipc> Drop for MultiBackend<T> {
+    fn drop(&mut self) {
+        // clear the select
+        std::mem::replace(&mut self.sel, Select::new());
+        // recover the box, but don't drop it just yet so we can clear the slice first
+        let _b = unsafe { Box::from_raw(self.receivers_ptr) };
+        // clear the slice
+        std::mem::replace(&mut self.receivers, &[]);
+        // now we can drop the box safely
+    }
+}
 
-    pub fn sender(&self) -> BackendSender<T> {
-        BackendSender(Rc::downgrade(&self.sock))
+impl<T: Ipc> Backend<T> for MultiBackend<T> {
+    fn sender(&self) -> BackendSender<T> {
+        match self.last_recvd {
+            Some(i) => self.backends[i as usize].clone(),
+            None    => panic!("Called sender but no messages have been received yet!")
+        }
     }
 
     /// Return a copy of the flag variable that indicates that the
     /// `Backend` should continue listening (i.e., not exit).
-    pub fn clone_atomic_bool(&self) -> Arc<atomic::AtomicBool> {
+    fn clone_atomic_bool(&self) -> Arc<atomic::AtomicBool> {
+        Arc::clone(&(self.continue_listening))
+    }
+
+    fn next(&mut self) -> Option<Msg> {
+        let oper = self.sel.select();
+        let index = oper.index();
+        self.last_recvd = Some(index);
+        oper.recv(&self.receivers[index]).unwrap() // TODO don't just unwrap
+    }
+}
+
+
+/// Backend will yield incoming IPC messages forever via `next()`.
+/// It owns the socket; `BackendSender` holds weak references.
+/// The atomic bool is a way to stop iterating.
+pub struct SingleBackend<T: Ipc> {
```

Comment:
Moving this up above `MultiBackend`would clean up the diff

Reply:
ack